mainloop: io_uring support

brauner opened the pull request:
Users can choose to compile liblxc with io_uring support. This will
cause LXC to use io_uring instead of epoll.
We're using both, io_uring's one-shot and multi-shot poll mode depending
on the type of handler.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

stgraber commented:
@brauner looks like this is break lxc-test-concurrent (timing out in Github Actions)

brauner commented:
> @brauner looks like this is break lxc-test-concurrent (timing out in Github Actions)

Yeah, this isn't done yet. I converted it from draft to review since I had a working version based on io_uring oneshot mode. But now I added support for multishot and will allow to mix and match oneshot and multishot mode depending on the handler.
However, this will need a revamp of the cleanup logic that we currently use in our event loops. Which is what I'm doing. I also need to make sure to keep our epoll implementation working of course. :)

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed
